common/datavalues/src/columns/data_column.rs

```diff
@@ -210,6 +261,135 @@ impl DataColumn {
+
+        let validity_to_apply = self.merge_validities(validities_to_apply)?;
+
+        // 1. If the validity to apply is all valid, no need to do the masking.
+        if validity_to_apply.all_valid() {
+            return Ok(self);
+        }
+
+        // 2. If the validity to apply is all null, just need to return an constant type with null value.
+        if validity_to_apply.all_null() {
+            let data_type = self.data_type();
+            let null_value = DataValue::new_from_data_type(&data_type, true);
+            let column = DataColumn::Constant(null_value, self.len());
+            return Ok(column);
+        }
+
+        // 3. Get the bitmap and do the masking.
+        let bitmap_to_apply = match validity_to_apply {
+            DataColumnValidity::Array(v, _) => v,
+
+            // The validity is neither all-true nor all-false, it must NOT be a constant one.
+            DataColumnValidity::Constant(_, _) => unreachable!(),
+        };
+
+        match self {
+            DataColumn::Array(_) => {
+                let array_ref = self.get_array_ref()?;
+                let current_bitmap: Option<&Bitmap> = array_ref.validity();
+                let bitmap_mask: Option<&Bitmap> = match &bitmap_to_apply {
+                    Some(bitmap) => Some(bitmap),
+                    None => None,
+                };
+                let new_bitmap = combine_validities(current_bitmap, bitmap_mask);
+                let array = self.get_array_ref()?.with_validity(new_bitmap);
+                let array_ref: ArrayRef = Arc::from(array);
+                let col = DataColumn::Array(array_ref.into_series());
+                Ok(col)
+            }
+            DataColumn::Constant(ref data_value, _) => {
+                // If the current data column is all constant null, no need to apply any masking.
+                if data_value.is_null() {
+                    return Ok(self);
+                }
+                let array = self.get_array_ref()?.with_validity(bitmap_to_apply);
+                let array_ref: ArrayRef = Arc::from(array);
+                let col = DataColumn::Array(array_ref.into_series());
+                Ok(col)
+            }
+        }
+    }
+
+    fn merge_validities(&self, validities: &[DataColumnValidity]) -> Result<DataColumnValidity> {
+        let array_len = self.len();
+
+        let mut bitmap_vec = vec![];
+
+        // Keep the index of the validity that is all null.
+        let mut all_null_index: i32 = -1;
+
+        for (i, v) in validities.iter().enumerate() {
+            if v.len() != array_len {
+                return Err(ErrorCode::BadArguments(format!(
+                    "Validity's length {} doesn't match column array len {}",
+                    v.len(),
+                    array_len
+                )));
+            }
+            if v.all_null() {
+                all_null_index = i as i32;
+            }
+            if !v.is_constant() {
+                bitmap_vec.push(v);
```

Comment:
So bitmap_vec can be Vec<Option<&Bitmap>> instead.  We don't need `unreachable` later.